lib/connection/result/row_stream.js

```diff
@@ -173,7 +173,37 @@ function RowStream(statement, context, options)
     // add the next row to the read queue
     process.nextTick(function ()
     {
-      self.push(externalizeRow(row, columns, mapColumnIdToExtractFnName));
+      // check if there are still rows available in the rowBuffer
+      if (rowBuffer && rowIndex > 0)
+      {
+        rowIndex--; // decrement the index to include the previous row in the while loop
+
+        // push() data to readable stream until highWaterMark threshold is reached or all rows are pushed
+        while (rowIndex < rowBuffer.length)
+        {
+          row = rowBuffer[rowIndex++];
+
+          // if buffer has reached the threshold based on the highWaterMark value then
+          // push() will return false and pause sending data to the buffer until the data is read from the buffer
+          if (!self.push(externalizeRow(row, columns, mapColumnIdToExtractFnName)))
+          {
+            break;
+          }
+        }
+
+        // check if all rows in rowBuffer has been pushed to the readable stream
+        if (rowIndex === rowBuffer.length)
+        {
+          // reset the buffer and index
+          rowBuffer = null;
+          rowIndex = 0;
+        }
+      }
+      else // No more rows left in the buffer
+      {
+        // Push the last row in the buffer
+        self.push(externalizeRow(row, columns, mapColumnIdToExtractFnName));
```

Comment:
Why special code for last row? what are chances we keep adding last row multiple times?